drgn/helpers/common/prog.py

```diff
@@ -0,0 +1,257 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+"""
+Program Decorators
+------------------
+
+The ``drgn.helpers.common.prog`` module provides decorators to transparently
+use the :ref:`default-program`.
+"""
+
+import functools
+import inspect
+import typing
+
+from drgn import Object, Program, get_default_prog
+
+__all__ = (
+    "takes_object_or_program_or_default",
+    "takes_program_or_default",
+)
+
+# This would require Python 3.10, but we don't need any of it at runtime.
+if typing.TYPE_CHECKING:
+    from typing import (  # novermin
+        Any,
+        Callable,
+        Concatenate,
+        Optional,
+        ParamSpec,
+        Protocol,
+        TypeVar,
+        overload,
+    )
+
+    P = ParamSpec("P")
+    R = TypeVar("R")
+    R_co = TypeVar("R_co", covariant=True)
+
+    class TakesProgram(Protocol[P, R_co]):
+        def __call__(self, prog: Program, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+    class TakesProgramOrDefault(Protocol[P, R_co]):
+        @overload
+        def __call__(self, prog: Program, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+        @overload
+        def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+    class TakesObjectOrProgramOrDefault(Protocol[P, R_co]):
+        @overload
+        def __call__(self, prog: Program, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+        @overload
+        def __call__(self, __obj: Object, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+        @overload
+        def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+
+def takes_program_or_default(f: "TakesProgram[P, R]") -> "TakesProgramOrDefault[P, R]":
+    """
+    Wrap a function taking a :class:`~drgn.Program` so that it uses the
+    :ref:`default-program` if omitted.
```

Comment:
Reading this docstring, it primed me to believe that this function would never use `Object.prog_`. Due to the language "it uses the default program if omitted" - in my reading, that means that there are two cases: either a program is provided by the user, or the default program is used.

In the implementation (and one example you give), that expectation is not correct - there's the third case: the user provides an object, in which case you use `Object.prog_`. I think it would be nice to have a second sentence in here that says something along the lines of `If the first argument is a :class:`~drgn.Object`, then its associated program will be used rather than the default program.`

Reply:
Yeah, I'm overloading the term default program in a few places to either mean "the program set by set_default_prog()" or "the program used if one is not explicitly given". I'll try to clarify that. Maybe the latter could be referred to as the "default program rules", or I can say the "program is automatically determined if omitted", linking to the rules.